File: JobDescription/src/src/repository/BaseFileRepository.java
package repository;

import domain.Entity;
import domain.Job;
import validator.Validator;

import java.io.*;

/**
 * Created by claudiu on 21.10.2016.
 */
public class BaseFileRepository<E extends Entity> extends BaseRepository<E> implements FileRepository<E> {

    private FileRepository<E> repository;

    public BaseFileRepository() {
    }

    public BaseFileRepository(FileRepository<E> repository) {
        super(repository.getValidator());
        this.repository = repository;
        this.loadData();
    }

    public void loadData() {
        try (BufferedReader reader = new BufferedReader(new FileReader(this.repository.getFileName()))) {
            String line;
            while ((line = reader.readLine()) != null) {
                E obj = this.createFromFormat(line);
                this.validator.validate(obj);
                this.add(obj);
            }
        } catch (Exception e) {
            e.printStackTrace();        }
    }

    public void writeData() {
        try (BufferedWriter writer = new BufferedWriter(new FileWriter(this.repository.getFileName()))) {
            for (E entity : this.getAll()) {
                writer.write(this.repository.transform(entity));
                writer.newLine();
            }
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    public E createFromFormat(String line) {
        return this.repository.createFromFormat(line);
    }

    public String transform(E entity) {
        return this.repository.transform(entity);
    }

    public String getFileName() {
        return this.repository.getFileName();
    }

    public Validator<E> getValidator() {
        return this.repository.getValidator();
    }

    public void finalize() {
        this.writeData();
    }
}

File: JobDescriptionCsharp/JobDescriptionCsharp/repository/BaseFileRepository.cs
using JobDescriptionCsharp.domain;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using JobDescriptionCsharp.validator;
using System.IO;

namespace JobDescriptionCsharp.repository
{
    public class BaseFileRepository<GenericType> : BaseRepository<GenericType>, IFileRepository<GenericType>
        where GenericType : Entity
    {
        protected IFileRepository<GenericType> repository;
        public string fileName { get; set; }

        public BaseFileRepository(IFileRepository<GenericType> repository) : base(repository.getValidator())
        {
            this.repository = repository;
            this.fileName = repository.getFileName();
            loadData();
        }

        public void loadData()
        {
            //@todo: change the search the path of file name
            using (StreamReader reader = new StreamReader(fileName))
            {
                string line;
                while ((line = reader.ReadLine()) != null)
                {
                    GenericType entity = createFromLine(line);
                    this.add(entity);
                }
            }
        }

        public virtual GenericType createFromLine(string line)
        {
            return repository.createFromLine(line);
        }

        public void writeData()
        {
            using (StreamWriter writer = new StreamWriter(@"E:\\Facultate\anul2\\MAP\\JobDescriptionCsharp\\JobDescriptionCsharp\\data\\s.txt"))
            {
                writer.WriteLine(transform(repository.getEntityById(1)));

                //repository.getAll().ForEach(delegate (GenericType item)
                //{
                //    writer.WriteLine(transform(item));
                //});
            }
        }

        public virtual string transform(GenericType item)
        {
            return repository.transform(item);
        }

        public IValidator<GenericType> getValidator()
        {
            return this.validator;
        }

        public string getFileName()
        {
            return this.fileName;
        }
    }
}
